### PR TITLE
New RotateSource algorithm

### LINKED ISSUE #14684: New RotateSource algorithm
Write an algorithm that rotates the source, given an angle of rotation and the rotation axis.

### PR DESCRIPTION
**Pull request by raquelalvarezbanos**
Fixes #14684 

We need this algorithm to fix issue #14575. The new algorithm basically moves the source to the correct position given the incident theta, that is, it performs a rotation around an axis centered at the sample's position, this axis is perpendicular to the plane described by the beam and `up` vectors, both read from the instrument reference frame.

To test: I am not sure if this is the most general approach, I am assuming for instance that we always want to rotate the source around the sample, and that the rotation axis is perpendicular to the beam and the `up` vectors. Code review should be enough. Alternatively, if this algorithm is used in `ReflectometryReductionOne` before `ConvertUnits`, the correct `IvsQ` should be produced.

Release notes: not sure if we need to mention this in the release notes?

### CODE REVIEW

**Comment by OwenArnold** [comment]
In answer to your questions. (1) Yes, I think we do want to rotate around the source, but we should document this somewhere. (2) Yes we do want to add this to the reflectometry release notes, but maybe after we are done with the fix.

File: Framework/DataHandling/inc/MantidDataHandling/RotateSource.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
+#ifndef MANTID_DATAHANDLING_ROTATESOURCE_H_
+#define MANTID_DATAHANDLING_ROTATESOURCE_H_
+
+#include "MantidDataHandling/DllConfig.h"
+#include "MantidAPI/Algorithm.h"
+namespace Mantid {
+namespace DataHandling {
+
+/** RotateSource : Moves the source by a given angle
```
Comment (OwenArnold) [inline Framework/DataHandling/inc/MantidDataHandling/RotateSource.h]: Should state that the centre of the rotation is the sample in the comment.

File: Framework/DataHandling/src/RotateSource.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+using namespace API;
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void RotateSource::init() {
+
+  // When used as a Child Algorithm the workspace name is not used - hence the
+  // "Anonymous" to satisfy the validator
+  declareProperty(new WorkspaceProperty<Workspace>("Workspace", "Anonymous",
```
Comment (OwenArnold) [inline Framework/DataHandling/src/RotateSource.cpp]: I think the input workspace should be a type of MatrixWorkspace? I don't think we'd want to allow rotation of a TableWorkspace for instance.

File: Framework/DataHandling/src/RotateSource.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                                                   Direction::InOut),
+                  "The name of the workspace for which the new instrument "
+                  "configuration will have an effect. Any other workspaces "
+                  "stored in the analysis data service will be unaffected.");
+  declareProperty("Angle", 0.0, "The angle of rotation in degrees.");
+}
```
Comment (OwenArnold) [inline Framework/DataHandling/src/RotateSource.cpp]: Should we specify which direction a +ve vs -ve means in terms of the angle?

File: Framework/DataHandling/src/RotateSource.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  MatrixWorkspace_sptr inputW =
+      boost::dynamic_pointer_cast<MatrixWorkspace>(ws);
+  DataObjects::PeaksWorkspace_sptr inputP =
+      boost::dynamic_pointer_cast<DataObjects::PeaksWorkspace>(ws);
+
+  // Get some stuff from the input workspace
+  Instrument_const_sptr inst;
+  if (inputW) {
+    inst = inputW->getInstrument();
+    if (!inst)
```
Comment (OwenArnold) [inline Framework/DataHandling/src/RotateSource.cpp]: Good checks

File: Framework/DataHandling/src/RotateSource.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  if (angle != 0.) {
+
+    // Need the reference frame to decide around which axis to rotate
+    auto refFrame = inst->getReferenceFrame();
+
+    if (!refFrame) {
+      throw std::runtime_error("Could not get a valid reference frame");
+    }
+
+    auto pointingAlong = refFrame->pointingHorizontal();
```
Comment (OwenArnold) [inline Framework/DataHandling/src/RotateSource.cpp]: You don't need to do it like this. You should just be able to use `ReferenceFrame::vecPointingAlongBeam` etc which should be a unit vector.

File: Framework/DataHandling/src/RotateSource.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    } else if (pointingAlong == Y) {
+      y = 1.;
+    } else if (pointingAlong == Z) {
+      z = 1.;
+    } else {
+      throw std::runtime_error("Could not get a valid rotation axis");
+    }
+
+    // The handedness
+    auto handedness = refFrame->getHandedness();
```
Comment (OwenArnold) [inline Framework/DataHandling/src/RotateSource.cpp]: Ok. Answers one of my questions above. We should document this.

File: Framework/DataHandling/test/RotateSourceTest.h [hunk 7]
```diff
@@ -0,0 +1,189 @@
+#ifndef MANTID_DATAHANDLING_ROTATESOURCETEST_H_
+#define MANTID_DATAHANDLING_ROTATESOURCETEST_H_
+
```
Comment (OwenArnold) [inline Framework/DataHandling/test/RotateSourceTest.h]: Really excellent testing!

File: docs/source/algorithms/RotateSource-v1.rst [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+.. alias::
+
+.. properties::
+
+Description
+-----------
+
+This algorithm corrects the source's position by rotating it around an axis centered at the sample.
+The rotation axis is perpendicular to the plane determined by the beam direction and the *up* direction.
```
Comment (OwenArnold) [inline docs/source/algorithms/RotateSource-v1.rst]: Ok. We just want to be explicit about how the handedness interplays with the angle sign here.

**Comment by OwenArnold** [comment]
Other than that. I've had a play with this, and it works as I would expect it to.

**Comment by raquelalvarezbanos** [comment]
Thanks for testing this @OwenArnold, I've added the suggested changes.

**Comment by OwenArnold** [comment]
Good work. Please update release notes.

**Comment by raquelalvarezbanos** [comment]
release notes updated: http://www.mantidproject.org/index.php?title=Release_Notes_3_6_Reflectometry&diff=25724&oldid=25601

**Comment by OwenArnold** [comment]
@raquelalvarezbanos Can you make sure as part of a separate issue (1) the Reduction algorithms perform this correction. (2) The workflow graphviz gets updated to show the `RotateSource`. I suggest that we only `RotateSource` if we are also *CorrectPosition*. What do you think?

**Comment by raquelalvarezbanos** [comment]
Yes, sounds reasonable, I agree it makes sense to perform the rotation only if we `CorrectPosition`. There's already an issue to address (1): #14575. I'll include (2) as part of the work.

**Pull request merged**